sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsGdsStore.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.io.File
+import java.util.concurrent.ConcurrentHashMap
+
+import ai.rapids.cudf._
+import com.nvidia.spark.rapids.StorageTier.StorageTier
+import com.nvidia.spark.rapids.format.TableMeta
+
+import org.apache.spark.sql.rapids.RapidsDiskBlockManager
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/** A buffer store using GPUDirect Storage (GDS). */
+class RapidsGdsStore(
+    diskBlockManager: RapidsDiskBlockManager,
+    catalog: RapidsBufferCatalog = RapidsBufferCatalog.singleton)
+    extends RapidsBufferStore("gds", catalog) with Arm {
+  private[this] val sharedBufferFiles = new ConcurrentHashMap[RapidsBufferId, File]
+
+  override def createBuffer(
+      other: RapidsBuffer,
+      stream: Cuda.Stream): RapidsBufferBase = {
+    withResource(other.getMemoryBuffer) { otherBuffer =>
+      val deviceBuffer = otherBuffer match {
+        case d: DeviceMemoryBuffer => d
+        case _ => throw new IllegalStateException("copying from buffer without device memory")
+      }
+      val id = other.id
+      val path = if (id.canShareDiskPaths) {
+        sharedBufferFiles.computeIfAbsent(id, _ => id.getDiskPath(diskBlockManager))
+      } else {
+        id.getDiskPath(diskBlockManager)
+      }
+      // When sharing files, append to the file; otherwise, write from the beginning.
+      val fileOffset = if (id.canShareDiskPaths) {
+        // only one writer at a time for now when using shared files
+        path.synchronized {
+          CuFile.copyDeviceBufferToFile(path, deviceBuffer, true)
+        }
+      } else {
+        CuFile.copyDeviceBufferToFile(path, deviceBuffer, false)
+      }
+      logDebug(s"Spilled to $path $fileOffset:${other.size} via GDS")
+      new RapidsGdsBuffer(id, fileOffset, other.size, other.meta, other.getSpillPriority)
+    }
+  }
+
+  class RapidsGdsBuffer(
+      id: RapidsBufferId,
+      fileOffset: Long,
+      size: Long,
+      meta: TableMeta,
+      spillPriority: Long) extends RapidsBufferBase(id, size, meta, spillPriority) {
+    private[this] var deviceBuffer: Option[DeviceMemoryBuffer] = None
+
+    override val storageTier: StorageTier = StorageTier.GDS
+
+    private def getDeviceBuffer: DeviceMemoryBuffer = synchronized {
+      if (deviceBuffer.isEmpty) {
```

Comment:
Trying to cache the GDS data in GPU memory via a GDS store buffer seems problematic to me.  What happens if GPU memory pressure pushes out all of the device memory buffers into the GDS store, then these buffers get referenced back into GPU memory?  When the next OOM occurs, none of these GDS buffers in GPU memory are spillable, as they're not tracked in the device memory store.

If we want to cache these buffers in GPU memory then I think we need to implement reverse-spill logic where these cached buffers are tracked in the device store.  Then when we try to spill from the device store we can prioritize "spilling" buffers that have already been spilled to GDS and trivially release them.  That is much more involved, so probably the first step is to copy the data back into a temporary, non-cached buffer.  Not great for handling spilled broadcast data, but should work OK for the typical case of spilled shuffle partitions.  We can implement the reverse-spill logic in a separate PR.

Reply:
Done.